transactions/WFLY-17742_SupportTransactionsDuringGracefulShutdown

```diff
@@ -0,0 +1,146 @@
+= Support Transactions during WildFly’s graceful shutdown
+:author:        	Manuel Finelli
+:email:         	jfinelli@redhat.com
+:toc:           	left
+:icons:         	font
+:idprefix:
+:idseparator:   	-
+
+== Overview
+
+When employing WildFly in cloud environments, distributed transactions are not supported out of the box. In particular, Narayana (and, thus, the Transactions subsystem) has some [hard requirements](https://jbossts.blogspot.com/2022/04/narayana-on-cloud-part-1.html) that need to be satisfied. The goal of this proposal is to outline how distributed transactions should be safely terminated during the graceful shutdown of WildFly, with a particular focus on cloud environments.
+
+=== Current Issues
+
+_1.1 Currently, WildFly’s graceful shutdown does not take into consideration that hosting entities can be shut down indefinitely, i.e. there will not be any restart/resume after shutdown/suspension. From the point of view of the Transactions subsystem, this scenario might develop into a data integrity issue. In fact, Narayana’s Recovery Coordinator does not necessarily resolve all in-doubt transactions during its suspension, leaving the resolution of left-over transactions for later, when the Recovery Coordinator gets resumed._
+
+The main purpose of the Recovery Coordinator is to recover transactions through a two-phase cycle. Narayana’s behaviour is premised on the assumption that the Recovery Coordinator will be suspended and then will eventually be resumed. During suspension, the Recovery Coordinator is instructed to attempt one extra recovery cycle before leaving all in-doubt transactions in the Object Store. When the Recovery Coordinator is resumed, the recovery routine starts again and it will try to take care of all in-doubt transactions found in the Object Store.
+It is the responsibility of the third-party integrating Narayana to handle how and when the Recovery Coordinator should be suspended. What is missing in Narayana is a coherent API that can be used to discover the number of in-flight and in-doubt transactions that are still due to be completed. When it comes to WildFly and its graceful shutdown (AKA suspension), the Transactions subsystem gets suspended calling directly the suspension hook of Narayana’s Recovery Coordinator, without considering if the transaction recovery ability is still needed (e.g. by other subsystems or customers’ applications running in the server). As discussed previously, this current configuration is fine in all cases where WildFly is resumed/restarted after suspension/shutdown but a potential data integrity issue can happen when WildFly is shut down forever.
```

Comment:
Initially I was thinking to change "coherent API" to something like a "public API". But after some research I did discover https://github.com/jbosstm/narayana/blob/main/ArjunaCore/arjuna/classes/com/arjuna/ats/arjuna/coordinator/TxStats.java#L123 and there is https://github.com/jbosstm/narayana/blob/main/ArjunaCore/arjuna/classes/com/arjuna/ats/arjuna/coordinator/TxStats.java#L123t - I am not sure it's a good match or not but FYI anyway. Whether these would do what is necessary or not, maybe you can still expand on what is intended by the use of "coherent" so that the requirement is clarified, please?

Reply:
> I'll either change it with something else ("public" can be a valid replacement, thank you Tom :-) ) or I'll expand what I intended with 'coherent'.

or both :-)

Reply:
I know this conversation is resolved but this work needs to go into an SPI, ie it's not for general consumption, it's for integrators only.

Reply:
Thanks for pointing this out.